src/Arrays/MPIStateArrays.jl

```diff
@@ -420,49 +382,127 @@ function finish_ghost_send!(Q::MPIStateArray)
-        wait(device(buf), event)
     end
 end
 
-function transferrecvbuf!(buf, recvbuf, vmaprecv)
-    if length(vmaprecv) > 0
-        Np = size(buf, 1)
-        nvar = size(buf, 2)
+function __Isend!(Q)
+    nnabr = length(Q.nabrtorank)
+    transfer = get_transfer(Q.send_buffer)
 
-        event = Event(device(buf))
-        event = knl_transferrecvbuf!(device(buf), 256)(
-            Val(Np),
-            Val(nvar),
-            buf,
-            recvbuf,
-            vmaprecv,
-            length(vmaprecv);
-            ndrange = length(vmaprecv),
-            dependencies = (event,),
+    for n in 1:nnabr
+        Q.sendreq[n] = MPI.Isend(
+            (@view transfer[:, Q.nabrtovmapsend[n]]),
+            Q.nabrtorank[n],
+            666,
+            Q.mpicomm,
         )
-        wait(device(buf), event)
     end
 end
 
+# {{{ MPI Buffer handling
+function fillsendbuf!(sendbuf, buf, vmapsend; dependencies = nothing)
+    if vmapsend == 0
+        return MultiEvent(dependencies)
+    end
+
+    Np = size(buf, 1)
+    nvar = size(buf, 2)
+
+    event = knl_fillsendbuf!(device(buf), 256)(
+        Val(Np),
+        Val(nvar),
+        sendbuf,
+        buf,
+        vmapsend,
+        length(vmapsend);
+        ndrange = length(vmapsend),
+        dependencies = dependencies,
+    )
+
+    return event
+end
+
+function transferrecvbuf!(buf, recvbuf, vmaprecv; dependencies = nothing)
+    if vmaprecv == 0
+        return MultiEvent(dependencies)
+    end
+
+    Np = size(buf, 1)
+    nvar = size(buf, 2)
+
+    event = knl_transferrecvbuf!(device(buf), 256)(
+        Val(Np),
+        Val(nvar),
+        buf,
+        recvbuf,
+        vmaprecv,
+        length(vmaprecv);
+        ndrange = length(vmaprecv),
+        dependencies = dependencies,
+    )
+
+    return event
+end
+
 # }}}
 
+"""
+    ghost_exchange!(Q::MPIStateArray; dependencies=nothing)
+
+Asynchronous call to perform the exchange of ghost degrees-of-freedom.  The
+returned `KernelAbstractions.Event` can be waited on to know when the ghost
+exchange is finished.
+"""
+function ghost_exchange!(Q::MPIStateArray; dependencies = nothing)
+    event = Event(__Irecv!, Q; dependencies = dependencies)
+    event = Event(__testall!, Q.recvreq; dependencies = event)
+    event = prepare_stage!(Q.recv_buffer; dependencies = event)
+    event = transferrecvbuf!(
+        Q.data,
+        get_stage(Q.recv_buffer),
+        Q.vmaprecv;
+        dependencies = event,
+    )
+    return_event = event
+
+    # Wait on previous sends
+    event = Event(__testall!, Q.sendreq; dependencies = dependencies)
+    event = fillsendbuf!(
+        get_stage(Q.send_buffer),
+        Q.data,
+        Q.vmapsend;
+        dependencies = event,
+    )
+    event = prepare_transfer!(Q.send_buffer; dependencies = event)
+    # Here we don't wait on the send using a `KernelAbstractions.Event`.  The
+    # wait happens above at the MPI level via `MPI.Testall!`.
+    Event(__Isend!, Q; dependencies = event)
```

Comment:
Dangling dependency edges like this make me worry. If nothing in the DAG waits on this there is no pressure to execute this task. So I would rather return a `MultiEvent` containing the `send` and `recv` event. Or return a tuple and thread this back into the dependency chain.